kpatch-build/create-diff-object.c

```diff
@@ -1340,6 +1340,9 @@ static void kpatch_correlate_static_local_variables(struct kpatch_elf *base,
 					target_sec = parent->sec->rela;
 			}
 
+			if (is_special_static(target_sec->base->secsym))
+				continue;
+
```

Comment:
I'm still confused by this, can you give a more specific example of why it's needed?

Reply:
With macro-callback.patch I am hitting diff_fatal on line 1347, because target_sec->twin is null. The section is '__dyndbg' (thats why it has no twin), the symbol is 'L.str'

Reply:
`.L.str` isn't actually a static local variable (static local means function scope).  kpatch_is_normal_static_local() is getting confused by the "." in the variable name.

Also, looking at a clang binary, static locals are named differently compared to GCC: they have the function name prepended to them, like:

```
    13: 0000000000000007     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.13
    14: 0000000000000009     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.17
    15: 0000000000000008     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.15
    16: 000000000000000a     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.19
    17: 000000000000000b     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.21
    18: 0000000000000005     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.9
    19: 0000000000000004     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.7
    20: 0000000000000006     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.11
    21: 0000000000000003     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once
    22: 000000000000000c     1 OBJECT  LOCAL  DEFAULT   40 unwind_next_frame.__print_once.23
```

This explains the need for the other __UNIQUE_ID_ change: those were also static locals.

Also notice that in clang, we no longer can rely on static locals having a numbered suffix.  In fact they usually *don't* have a numbered suffix, unless there's more than one of them (with the same name) in a function.

However, we *can* rely on them having a "." in the name, thanks to the function name being prepended.

So looking at the code I think we need some tweaks:

- (not clang-specific, just something I found as shown above) `__print_once` should be added to the special static list.  It's similar to `__warned` in functionality.

- is_special_static() needs to look for both GCC and Clang variants, e.g.  `__print_once.` (for GCC) and `.__print_once` (for Clang).  The prefix searching logic can use strstr() instead.

- kpatch_is_normal_static_local() should consider a ".L" symbol to be *not* a static local

- kpatch_mangled_strcmp() needs to consider `unwind_next_frame.__print_once` and `unwind_next_frame.__print_once.1` as equal, even though the former doesn't have a period at the end.

- it would probably be a good idea to add a unit test or two in this area

Reply:
Also while you're changing that code, I think kpatch_is_normal_static_local() and is_special_static() should really return bool :-)

Reply:
Thank you for this explaination. I've covered all of the mentioned changes in my last push. I am having trouble with unit tests though. Any hints on how I can get create-diff-object fail without a proper clang-style special static detection in is_special_static()?

Reply:
You could make a patch which removes one of the special statics.  Before your patch it thinks they're just normal static locals, so it tries to correlate them and errors out when one of them goes missing.  For example, looking at the vmlinux.o symbol table for ideas:
```
 25232: 0000000000000000  1246 FUNC    LOCAL  DEFAULT 18984 init_srcu_struct_fields
 25233: 0000000000000000     0 OBJECT  LOCAL  DEFAULT 19034 init_srcu_struct_fields.__key
 25234: 0000000000000000     0 OBJECT  LOCAL  DEFAULT 19035 init_srcu_struct_fields.__key.7
```
init_srcu_struct_fields() has three static locals named `__key` (from the mutex_init macro).  So remove one of those mutex_init() calls.

Reply:
Including `__print_once` in special static list broke `gcc-static-local-var-2.patch` on rhel8.1 and 8.2 on ppc64le. There are 2 `__print_once` symbols that are placed in `.data..read_mostly` section which is not in the list of data sections allowed for inclusion.  I thought maybe this can be treated the same way as .data.once but it doesn't seem to be the case as it contains other symbols which are not selected for inclusion so create-diff-object fails in `kpatch_create_intermediate_sections()`->`need_dynrela()` trying to work with symbols that were already torn-down.

Reply:
Instead of checking the variable name for `__print_once`, maybe just check if the symbol lives in `.data.once`.

Reply:
Implemented by checking sym->sec->name in is_special_static(): https://github.com/dynup/kpatch/pull/1156/files#diff-b11891a551552359c664b4475a7ed9f84b1d569d4c649a775c9ba23e53f4bbc8R347